### ISSUE TITLE
[Alerting V2 UI] Custom Component for Alertmanager Routes - Detail Page

### ISSUE THREAD

**Issue by aiyengar2** [rancher/dashboard#1239]
### Feature Request - Custom Component for Alertmanager Routes - Detail Page

This ticket specifically addresses creating a Detail Page UI in order to specify values that should be on the `routes` field of the [YAML-based configuration](https://prometheus.io/docs/alerting/latest/configuration/) specified within the `alertmanager.yaml` key within the Alertmanager Config Secret. A route represents the node on a routing tree that Alertmanager uses to determine which alerts get sent to particular receivers configured in https://github.com/rancher/dashboard/issues/1238.

See the context section below for more information on this.

#### How should the YAML look like for this object?

The YAML for the `routes` field (representing a tree of routes that would be configured on this Detail page) may look something like this:

```yaml
route:
  group_by: ['job']
  group_wait: 30s
  group_interval: 5m
  repeat_interval: 12h
  receiver: 'receiver-1'
  routes:
  - group_wait: 30s
    group_interval: 5m
    repeat_interval: 12h
    match:
      team: team-2
    receiver: 'receiver-2'
  - <more-routes>
```

The full spec for this object can be found [here](https://prometheus.io/docs/alerting/latest/configuration/#route).

#### Important: Rancher UI Constraints

1. Rancher will only be supporting this UI for a routing tree with two levels (i.e. non-root routes should not use the `routes` field to define children). If a user has an Alertmanager Secret that has a non-null / empty `route.routes[i].routes` for any `i`, we should automatically redirect the user to the corresponding Secret view / edit page to directly manage the Secret.
1. The routes that can be configured via this Detail page should only be routes found in `route.routes`. The master route (whose config is in `route` itself) should be configured on the List page of the routes.

#### What should the UI be for each Route on the Detail page?

The following UI should be exposed:

*Help Text:* 
```
Each route defines a rule on whether or not to send a notification to a configured notifier based on the labels of triggered alerts.
```
For the following table below, the k8s YAML prefix where the field is stored will be `.route.routes[i].*`:
Header / Label Name | k8s YAML | Description | Input | Default
---|---|---|---|---|
Receiver | `.receiver` | `Select from the list of configured receivers` | Ability to select Receivers by name (i.e. values in `.receivers[i].name` for all `I`) or select `No Receiver`, which is the default |  `'null'` (dropdown should show `No Receiver`)
Group By | `.group_by` | The labels by which alerts are grouped together | List of `string` | `[]`
Group Wait | `.group_wait` | How long to initially wait to send a notification for a group of alerts. | `string` | `30s`
Group Interval | `.group_interval` | How long to wait before sending a notification about new alerts that are added to a group of alerts for which an initial notification has already been sent. | `string` | `5m`
Repeat Interval | `.repeat_interval` | How long to wait before sending a notification again if it has already been sent successfully for an alert. | `string` | `4h`
Match | `route.match` | Alerts whose labels match these values will be sent by this receiver or processed by any configured child nodes | `key-value pairs` | `{}`
Match Regex | `route.match_re` | Alerts whose labels match the regex provided will be sent by this receiver or processed by any configured child nodes | `key-value pairs`, where the value is regex | `{}`

When a user creates / edits any of these routes, the corresponding object should be added / modified to the `route.routes` field of the `alertmanager.yaml` key on the Alertmanager Config Secret. Order must be preserved. Multiple routes can have the same `Receiver` field. On create only, `.continue` should automatically be set `true`. On submit, either the list provided for `Match` or `Match Regex` must be non-nil / non-empty and `Receiver` must be set.

When a user deletes any of these routes, the corresponding object should be removed from the `route.routes` field of the `alertmanager.yaml` key on the Alertmanager Config Secret. 

### Context

As part of the Alerting V2 chart deployment, we either create or use a single Secret in the `cattle-monitoring-system` namespace whose name is given by the Chart deployment value of `.alertmanager.alertmanagerSpec.configSecret`. This Secret contains the following keys:
- `alertmanager.yaml`: a required key. This represents the [YAML-based configuration](https://prometheus.io/docs/alerting/latest/configuration/) of the default Alertmanager instance deployed as part of Monitoring V2.
- one or more keys that correlate to template files that will be stored in `/etc/alertmanager/config/<key-name>` on the node. By default, the template file deployed by the chart deployment is `rancher_defaults.tmpl` but more templates could be specified by users; these templates can be used as values within the [YAML-based configuration](https://prometheus.io/docs/alerting/latest/configuration/) specified in the `alertmanager.yaml` key.

Specifically, we want to provide a way for users to easily configure the [YAML-based configuration](https://prometheus.io/docs/alerting/latest/configuration/) within the `alertmanager.yaml` key within the Secret. Once specified, the contents of the key will look something like this:

```yaml
global:
  resolve_timeout: 5m
route:
  group_by: ['job']
  group_wait: 30s
  group_interval: 5m
  repeat_interval: 12h
  receiver: 'null'
  routes:
  - group_by: ['job']
    group_wait: 30s
    group_interval: 5m
    repeat_interval: 12h
    match:
      alertname: Watchdog
    receiver: 'slack'
receivers:
- name: 'null'
- name: 'slack'
templates:
- /etc/alertmanager/config/*.tmpl
```

**Comment by lvuch** [comment]
https://xd.adobe.com/view/672f4601-a9de-4701-94fb-060f0d72667c-327c/

**Comment by codyrancher** [comment]
@deniseschannon @vincent99 @lvuch It's easy for me to switch to the mocks but I wanted to see if anyone preferred this tabbed version over the mocks. It felt more consistent with the other pages we've done to me.

![Screen Shot 2020-10-26 at 11 27 05 AM](https://user-images.githubusercontent.com/55104481/97213019-5c7dd280-177e-11eb-83ff-b015ae405676.png)
![Screen Shot 2020-10-26 at 11 27 14 AM](https://user-images.githubusercontent.com/55104481/97213015-5be53c00-177e-11eb-8347-a37cd74c9038.png)
![Screen Shot 2020-10-26 at 11 27 23 AM](https://user-images.githubusercontent.com/55104481/97213012-5ab40f00-177e-11eb-842c-7e4d53d58fd8.png)

**Comment by aiyengar2** [comment]
@codyrancher personally I prefer the mocks. 

I think the tab-based approach works great when the first tab contains all the basic / required configuration that needs to be done for any resource and the other tabs are "optional" or "advanced configuration" that don't have to be configured; that way, a user would not have to click through each tab unless they want that additional stuff.

However, in Routes both the `Receiver` and `Matching` sections' fields are required for any Route either way, so using the tabbed approach would require a user to click through each tab to save the resource.

Also, there isn't much configuration per tab, so I think the single page approach in the mockups looks good to me.

**Comment by aiyengar2** [comment]
@codyrancher as discussed offline, the `Ability to select Receivers by name (i.e. values in .receivers[i].name for all i)` should also include the ability to select `No Receiver`.

If `No Receiver` is selected, `route.receiver` or `route.routes[i].receiver` should be set to `'null'`, just like in the monitoring chart's default receiver: https://github.com/rancher/charts/blob/5621a9bfc985d479aaf496cdab951abf82c480fa/charts/rancher-monitoring/values.yaml#L404

**Comment by jiaqiluo** [comment]
on rancher:`v2.5-cc777ebfabe4dba48216525bb1cd56c01231dd40-head` with the builtin UI 

no `create` button available on the route page 

![image](https://user-images.githubusercontent.com/6218999/97759213-ccad9080-1abd-11eb-9c10-5e2d2159fb89.png)

**Comment by jiaqiluo** [comment]
### issue 1: invalid field `name` for routes
The field `name` does not exist for the route, we need to remove it because it breaks the configuration file.
Note: 
- the filed `name` does not show up when we edit the existing routes as YAML, but it shows up in the content of the secret` alertmanager-rancher-monitoring-alertmanager`  
- even worse, the routes are gone from the Dashboard after I remove the `name` from the secret. It looks like `name` is used by the UI to get/list the existing routes. 
```
route:
  routes:
    - group_by: []
      group_wait: 30s
      group_interval: 5m
      repeat_interval: 4h
      match:
        env: dev
      match_re: {}
      name: route-1    ----------> this line
      receiver: test-1
  receiver: test-1
  group_by:
    - env
  group_wait: 5s
  group_interval: 2m
  repeat_interval: 1m
```

**Comment by codyrancher** [comment]
@jiaqiluo I'm planning on keeping the name field as it shouldn't negatively impact monitoring. I'm changing how we deal with the lack of names though. 

If a name isn't present then we give it a default name of `route-{i}` where `i` is the position it's stored in the secret.

**Comment by jiaqiluo** [comment]
> I'm planning on keeping the name field as it shouldn't negatively impact monitoring.

It does affect monitoring, especially the alerting part. The alerting manager will fail to load/reload the configuration from the secret because of the name field, which means you cannot add any new alerts.

**Comment by jiaqiluo** [comment]
The fix is validated on rancher:`v2.5-2859fc5b7ae901c8bae22d7820df0598380142bb-head` with the latest UI

- issue 1: fixed. Now the name is filled by the UI and is not in the YAML file. The routes are created and loaded into the Alertmagner successfully. 

<img width="368" alt="Screen Shot 2020-11-04 at 10 22 55 AM" src="https://user-images.githubusercontent.com/6218999/98144987-e36e3180-1e87-11eb-945d-76f788fe9c3b.png">
